yjit/src/asm/mod.rs

```diff
@@ -276,6 +287,37 @@ impl CodeBlock {
         self.mem_block.mark_all_executable();
     }
 
+    /// Free unused code pages
+    fn code_gc(&mut self) {
+        let iseq_payloads = CodegenGlobals::get_iseq_payloads();
+
+        // Check which pages are still in use
+        let num_pages = self.mem_block.num_pages();
+        let mut alive_pages = vec![false; num_pages]; // TODO: consider using a bitmap
+        for iseq_payload in iseq_payloads.iter() {
+            let iseq_payload = unsafe { iseq_payload.as_ref() }.unwrap();
+            for (start_addr, end_addr) in iseq_payload.code_addrs() {
+                for page in self.mem_block.addrs_to_pages(start_addr, end_addr) {
+                    alive_pages[page] = true;
+                }
+            }
+        }
+
+        // Ask VirtualMem to free unused pages
+        for (page, alive) in alive_pages.iter().enumerate() {
+            if !alive {
+                // TODO: reduce the number of syscalls by merging them for consecutive pages
+                let freed_bytes = self.mem_block.free_page(page);
+                #[cfg(feature = "stats")]
+                {
+                    self.freed_bytes += freed_bytes;
+                }
+            }
+        }
+
+        // TODO: Implement start-over; allow reusing freed pages
```

Comment:
I guess we will need a list of free pages for this 🤔

Reply:
I'm thinking, if we use the reference count bitmap idea for incremental code GC, we could reuse it for finding which page is used by no ISEQs.

Reply:
Possibly but we probably don't want to have to iterate and compare lots of integers (up to ~16K integers) each time we want to find a free page.

Reply:
I don't think it needs to compare all the integers every time. It should have a moving pointer to a page and only scan pages until it finds an unused page. Good news is that you'd be able to compare 64 pages at once. If at least one page is unused in 64 pages, looking at a single 64-bit range of the bitmap would be enough.

Reply:
So you'd have a refcount for each page to detect when it becomes free, and also a bitmap to keep track of all free pages, and then an index into the bitmap to keep track of where to find the first free page (lowest index)?